pinot-plugins/pinot-stream-ingestion/pinot-rocketmq/src/main/java/org/apache/pinot/plugin/stream/rocketmq/RocketMQStreamLevelConsumer.java

```diff
@@ -0,0 +1,155 @@
+
+  private long _lastLogTime = 0;
+  private long _lastCount = 0;
+  private long _currentCount = 0L;
+
+  public RocketMQStreamLevelConsumer(String clientId, String tableName, StreamConfig streamConfig,
+      Set<String> sourceFields, String groupId) {
+    _streamConfig = streamConfig;
+    _rocketmqStreamLevelStreamConfig = new RocketMQConfig(streamConfig, groupId);
+
+    _messageDecoder = StreamDecoderProvider.create(streamConfig, sourceFields);
+
+    _instanceLogger =
+        LoggerFactory.getLogger(RocketMQConfig.class.getName() + "_" + tableName + "_" + streamConfig.getTopicName());
+    _instanceLogger.info("RocketMQStreamLevelConsumer: streamConfig : {}", _streamConfig);
+  }
+
+  @Override
+  public void start()
+      throws Exception {
+    _consumer = RocketMQStreamLevelConsumerManager.acquireRocketMQConsumerForConfig(_rocketmqStreamLevelStreamConfig);
+  }
+
+  private void updateRocketMQIterator() {
+    _consumerRecords = _consumer.poll(_streamConfig.getFetchTimeoutMillis());
+    _rocketmqIterator = _consumerRecords.iterator();
+  }
+
+  private void resetOffsets() {
+    for (MessageQueue queue : _consumerOffsets.keySet()) {
+      long offsetToSeek = _consumerOffsets.get(queue);
+      try {
+        _consumer.seek(queue, offsetToSeek);
+      } catch (MQClientException e) {
+        _instanceLogger.warn("Caught exception while seek offset", e);
+      }
+    }
+  }
+
+  @Override
+  public GenericRow next(GenericRow destination) {
+    if (_rocketmqIterator == null || !_rocketmqIterator.hasNext()) {
+      updateRocketMQIterator();
+    }
+    if (_rocketmqIterator.hasNext()) {
+      try {
+        final MessageExt record = _rocketmqIterator.next();
+        updateOffsets(new MessageQueue(record.getTopic(), record.getBrokerName(), record.getQueueId()),
+            record.getQueueOffset());
+        destination = _messageDecoder.decode(record.getBody(), destination);
+
+        _currentCount++;
+
+        final long now = System.currentTimeMillis();
+        // Log every minute or 100k events
+        if (now - _lastLogTime > 60000 || _currentCount - _lastCount >= 100000) {
+          if (_lastCount == 0) {
+            _instanceLogger
+                .info("Consumed {} events from rocketmq stream {}", _currentCount, _streamConfig.getTopicName());
+          } else {
+            _instanceLogger.info("Consumed {} events from rocketmq stream {} (rate:{}/s)", _currentCount - _lastCount,
+                _streamConfig.getTopicName(), (float) (_currentCount - _lastCount) * 1000 / (now - _lastLogTime));
+          }
+          _lastCount = _currentCount;
+          _lastLogTime = now;
+        }
+        return destination;
+      } catch (Exception e) {
+        _instanceLogger.warn("Caught exception while consuming events", e);
+        throw e;
+      }
+    }
+    return null;
+  }
+
+  private void updateOffsets(MessageQueue queue, long offset) {
+    _consumerOffsets.put(queue, offset + 1);
```

Comment:
Also, do we need to track the current consumed offset? It doesn't look like it is being "read" anywhere?

Reply:
Actually, it is used in [resetOffsets@82](https://github.com/apache/pinot/pull/8274/files/0df2ebcfd8891dbb7a1dbeca0f8fd5556e969ff5#diff-1ccff2998469b33edf9814080c3edcfe3bcd733189441a027ea94222246d1574R82).